### DISCUSSION TITLE
Code completion in GitHub Copilot for Xcode is now generally available

### DISCUSSION THREAD

**Discussion by Akash1134** [community/community#141898] [Copilot News and Announcements]
## 📢  Code completion in [GitHub Copilot for Xcode](https://aka.ms/copilotxcode) is now generally available 🎉 

<img alt="GitHub Copilot for Xcode code completion [GA]" src="https://github.com/user-attachments/assets/7d73ef6d-86e2-425f-bcba-3e7c4405bf87" width="800" />

Developers using Xcode can now experience AI-powered, real-time code suggestions that enhance productivity and streamline the development process.

### **Why You'll Love Copilot for Xcode** 💟 

- **Instant Inspiration**: Say goodbye to writer’s block! With real-time suggestions flowing as you type, you’ll be amazed at how quickly you can bring your ideas to life.
- **All the Languages, All the Fun**: Whether you're rocking Swift or diving into Objective-C, Copilot speaks your language. It’s like having a multilingual coding wizard by your side!
- **Unlock the Multiline Magic**: Stuck on a tricky piece of code? Hold down **Option + Tab** to reveal a treasure trove of multiline suggestions. Complex challenges? No problem!
- **Safe & Sound Coding**: With built-in content filters, you can code with confidence, knowing that all suggestions are clean and professional.
- **Stay Original**: Keep your code fresh! Our duplicate detection feature ensures that you won't receive suggestions that match public code, giving your projects that unique touch.

### **How to Join the Fun:**
1. **Grab Your Copilot License**: You’ll need one to embark on this adventure—available for all individual, business, and enterprise users.
2. **Install the Extension**: Check out our [getting started guide](https://docs.github.com/en/copilot/managing-copilot/configure-personal-settings/installing-the-github-copilot-extension-in-your-environment?tool=xcode) to get Copilot rolling in Xcode.
3. **Dive In and Experiment**: Don’t hold back! Try different features and let your creativity flow. What new coding shortcuts will you discover? How can Copilot make your workflow smoother?

### **We Want Your Feedback! 💬**
Once you’ve had a chance to explore, we’d love to hear from you! Share your thoughts, experiences, and any creative solutions you've discovered in the comments below or open up an issue on our [CopilotForXcode repo](https://github.com/github/CopilotForXcode/issues). Your feedback is invaluable in shaping the future of Copilot for Xcode!

### **Join the Adventure! 🌈**
So, what are you waiting for? Jump into the public preview of GitHub Copilot for Xcode and let the exploration begin! We can’t wait to see the incredible things you create with Copilot as your sidekick. Happy coding! 💻✨

**Comment by cjgaspari** [comment]
Looks awesome. One thing that seems to be missing is the "Ask Copilot" chat that Visual Studio Code has. Not sure if thats possible within the Xcode interface, but perhaps in the menubar Copilot icon, there could be a way to click into a chat view. 

Copilot code completion itself is amazing, however I really enjoy the conversational "Ask Copilot" in Visual Studio Code as more of a 'rubber duck' method of getting what I need -- in part to also understand the changes that are being made. 

Really excited to see all the work the team has done to get Copilot on Xcode!! Can't wait to see how this develops over time.

**Comment by nickjbedford** [comment]
Why exactly does the extension need to control my computer via Accessibility "control your computer" permission? Seems like an overreach and it's vague as to what it will even do.

I'm very wary of granting such permissions without understanding why they are needed.

**Reply by devm33** [reply]
Hi @nickjbedford, thanks for the question. That's a completely valid concern.

- The Accessibility permission is used to receive real-time updates to the current editor in Xcode. Xcode's [extension API](https://developer.apple.com/documentation/xcodekit/creating-a-source-editor-extension) is very limited and can only be manually invoked by the user.
- It's also used to interact with Xcode, e.g. to accept a suggestion.
- It's not used on any applications outside of Xcode, but unfortunately the permission is all or nothing -- there's no fine-grain scope.

Please feel free to audit the usage. The application is open source at https://github.com/github/CopilotForXcode You can search for `CGEvent` and `AX`* to see where the Accessibility permission is used.

**Reply by nickjbedford** [reply]
> Hi @nickjbedford, thanks for the question. That's a completely valid concern.
> 
> * The Accessibility permission is used to receive real-time updates to the current editor in Xcode. Xcode's [extension API](https://developer.apple.com/documentation/xcodekit/creating-a-source-editor-extension) is very limited and can only be manually invoked by the user.
> * It's also used to interact with Xcode, e.g. to accept a suggestion.
> * It's not used on any applications outside of Xcode, but unfortunately the permission is all or nothing -- there's no fine-grain scope.
> 
> Please feel free to audit the usage. The application is open source at https://github.com/github/CopilotForXcode You can search for `CGEvent` and `AX`* to see where the Accessibility permission is used.

Thanks for explaining, @devm33. Perhaps it may be worth adding some of this information to the settings application (or somewhere) to inform users of why the Accessibility permission is used as I'm sure developers would be more likely to question it than typical users. If I have time, I may look at the project myself as well.

**Reply by devm33** [reply]
💯 Absolutely a good idea. Thanks for the feedback!

**Comment by margolisdavid** [comment]
Getting a lot of "error language server timed out" errors. Otherwise love it so far. But need "Ask CoPilot" :)

**Reply by devm33** [reply]
Thanks for trying it out @margolisdavid! For the timeout errors, could you [open an issue](https://github.com/github/CopilotForXcode/issues/new?template=bug_report.md) to investigate. In particular, enabling verbose logs and attaching them to the issue would be helpful!

**Comment by FlorinDMass** [comment]
I can't use it because I have a root self-sign certificate.
Any idea how to solve it ?
PS: I created a bug here : https://github.com/github/CopilotForXcode/issues/29

**Reply by devm33** [reply]
Hi @FlorinDMass, thanks for trying out Copilot for Xcode and raising the issue. Will follow up on the bug!

**Comment by ggrell** [comment]
I'm running into an issue where the plugin/extension is showing code suggestions, but pressing Tab just inserts tab spaces, never adds the completion. It used to work for a bit. "Accept suggestions with Tab" is on. I've restarted Xcode (16.2) several times, also restarted GitHub Copilot app, but no luck.

If I choose Editor -> GitHub Copilot -> Accept Suggestion, then it is inserted. I've also tried disabling Xcode's built in completions and predictive suggestions.

**Reply by testforstephen** [reply]
@ggrell thanks for trying out. Not sure if some cache is broken. Could you try to delete `~/Library/Containers/com.github.CopilotForXcode.EditorExtension` and relaunch GitHub Copilot app? If it does not work, you might try to reinstall GitHub Copilot for Xcode.

**Reply by ggrell** [reply]
After a reinstall and a reboot, it seems to be working again!

**Comment by Mutix** [comment]


**Reply by lijie-lee** [reply]
Hi @Mutix, sorry for the confusion. The Readme has been updated without refering to the **Preview Policy**

**Comment by Bradleycorn** [comment]
When you open the chat window, which model is being used?  ChatGPT? Claude Sonnet? There doesn't appear to be any way to set the model that is used, nor any indication as to which model is being used.

**Reply by lijie-lee** [reply]
Hi @Bradleycorn. The [GitHub Copilot Chat for Xcode is Now Generally Available](https://github.com/orgs/community/discussions/151295) with **Model Selector** feature. Please upgrade to the laset version and have a try!
